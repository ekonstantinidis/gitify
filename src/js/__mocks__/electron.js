window.Notification = function (title) {
  this.title = title;

  return {
    onclick: jest.fn(),
  };
};

window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn(),
};

window.alert = jest.fn();

const browserWindow = {
  loadURL: jest.fn(),
  webContents: {
    on: () => {},
    session: {
      clearStorageData: jest.fn(),
    },
  },
  on: () => {},
  close: jest.fn(),
  hide: jest.fn(),
  destroy: jest.fn(),
};

const dialog = {
  showErrorBox: jest.fn(),
};

module.exports = {
  remote: {
    BrowserWindow: () => browserWindow,
    dialog: dialog,
    app: {
      getVersion: () => '0.0.1',
    },
    getCurrentWindow: jest.fn(() => browserWindow),
  },
  ipcRenderer: {
    send: jest.fn(),
    on: jest.fn(),
  },
  shell: {
    openExternal: jest.fn(),
  },
};
