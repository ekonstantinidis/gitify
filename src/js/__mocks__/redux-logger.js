export function createLogger() {
  return () => (next) => (action) => {
    return next(action);
  };
}
