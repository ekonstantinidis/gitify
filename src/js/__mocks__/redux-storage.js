export function createMiddleware() {
  return () => {
    return (next) => (action) => {
      return next(action);
    };
  };
}

export function createLoader() {
  const state = {
    auth: {
      token: 'LOGGED_IN_TOKEN',
    },
  };

  return () => new Promise((resolve) => resolve(state));
}

export function reducer(clb) {
  return (state, action) => clb(state, action);
}
