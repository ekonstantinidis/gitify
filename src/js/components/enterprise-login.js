var ipcRenderer = require('electron').ipcRenderer;

import React from 'react';
import helpers from '../utils/helpers';
import { connect } from 'react-redux';
import { updateSetting, loginUser } from '../actions';

class EnterpriseLogin extends React.Component {

  static contextTypes = {
    router: React.PropTypes.object.isRequired
  };

  componentWillReceiveProps(nextProps) {
    const isLoggedIn = nextProps.token !== null;
    debugger;
    if (isLoggedIn) {
      ipcRenderer.send('reopen-window');
      this.context.router.push('/notifications');
    }
  }

  componentDidMount() {
    // We want to know whether or not we are loading as enterprise
    this.props.updateSetting('isEnterprise', true);
  }

  toggleSetting(key, event) {
    this.props.updateSetting(key, event.target.value);
  }

  handleLoginClick = () => {
    helpers.authGithub(this.props.settings, this.props.loginUser);
  }

  routeToLogin = () => {
    this.context.router.push('/login');
  }

  render() {
    return (
      <div className="container-fluid main-container settings">
        <ul className="nav nav-pills">
          <li className="nav-item">
            <a className="nav-link" onClick={this.routeToLogin}>
              <i className="fa fa-arrow-left" /> Back to login page
            </a>
          </li>
        </ul>

        <div className="row setting">
          <div className="col-xs-4">Base Url</div>
          <div className="col-xs-8">
            <input
              className="form-control"
              placeholder="github.company.com"
              onChange={this.toggleSetting.bind(this, 'baseUrl')}
            />
          </div>
        </div>
        <div className="row setting">
          <div className="col-xs-4">Client ID</div>
          <div className="col-xs-8">
            <input
              className="form-control"
              placeholder="24084b7c891c8a7a0385"
              onChange={this.toggleSetting.bind(this, 'clientId')}
            />
          </div>
        </div>
        <div className="row setting">
          <div className="col-xs-4">Secret</div>
          <div className="col-xs-8">
            <input
              className="form-control"
              placeholder="177537vsrh5u54u2ht43y2yfhsr"
              onChange={this.toggleSetting.bind(this, 'clientSecret')}
            />
          </div>
        </div>

        <div className="row footer">
          <div className="col-xs-12 text-center">
            <button className="btn btn-md" onClick={this.handleLoginClick}>
              <i className="fa fa-github" />Login to GitHub enterprise
            </button>
          </div>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    token: state.auth.get('token'),
    settings: state.settings
  };
};

export default connect(mapStateToProps, {
  updateSetting,
  loginUser
})(EnterpriseLogin);
