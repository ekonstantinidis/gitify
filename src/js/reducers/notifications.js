import { Map, List } from 'immutable';
import {
  NOTIFICATIONS,
  MARK_NOTIFICATION,
  MARK_REPO_NOTIFICATION,
  MARK_ACCOUNT_NOTIFICATION,
  LOGOUT,
} from '../actions';

const initialState = Map({
  response: List(),
  isFetching: false,
  failed: false,
});

// Response Structure
// response: List([
//   Map({
//     hostname: 'github.com',
//     notifications: List([1, 2, 3])
//   }),
//   Map({
//     hostname: 'git-enterprise.hexwebs.com',
//     notifications: List([1, 2, 3])
//   }),
// ]);

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case NOTIFICATIONS.REQUEST:
      return state.set('isFetching', true).set('failed', false);
    case NOTIFICATIONS.SUCCESS:
      return state.set('isFetching', false).set('response', action.payload);
    case NOTIFICATIONS.FAILURE:
      return state
        .set('isFetching', false)
        .set('failed', true)
        .set('response', List());
    case MARK_NOTIFICATION.SUCCESS:
      const accNotificationsIndex = state
        .get('response')
        .findIndex(obj => obj.get('hostname') === action.meta.hostname);

      return state.updateIn(
        ['response', accNotificationsIndex, 'notifications'],
        notifications => {
          return notifications.filterNot(
            items => items.get('id') === action.meta.id
          );
        }
      );
    case MARK_REPO_NOTIFICATION.SUCCESS:
      const accNotificationsRepoIndex = state
        .get('response')
        .findIndex(obj => obj.get('hostname') === action.meta.hostname);

      return state.updateIn(
        ['response', accNotificationsRepoIndex, 'notifications'],
        notifications => {
          return notifications.filterNot(
            items =>
              items.getIn(['repository', 'full_name']) === action.meta.repoSlug
          );
        }
      );
    case MARK_ACCOUNT_NOTIFICATION.SUCCESS:
      const accNotificationsAccountIndex = state
        .get('response')
        .findIndex(obj => obj.get('hostname') === action.meta.hostname);

      return state.updateIn(
        ['response', accNotificationsAccountIndex, 'notifications'],
        notifications => notifications.clear()
      );
    case LOGOUT:
      return initialState;
    default:
      return state;
  }
}
