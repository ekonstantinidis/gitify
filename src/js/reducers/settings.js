import { UPDATE_SETTING, HAS_STARRED, TOGGLE_SETTINGS_MODAL } from '../actions';
import { Map } from 'immutable';

const initialState = Map({
  participating: false,
  playSound: true,
  showNotifications: true,
  markOnClick: false,
  openAtStartup: false,
  showSettingsModal: false,
  hasStarred: false,
  showAppIcon: 'both',
  isEnterprise: false,
  baseUrl: 'github.com',
  clientId: '3fef4433a29c6ad8f22c',
  clientSecret: '9670de733096c15322183ff17ed0fc8704050379'
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case UPDATE_SETTING:
      return state.set(action.setting, action.value);
    case HAS_STARRED.SUCCESS:
      return state.set('hasStarred', true);
    case HAS_STARRED.FAILURE:
      return state.set('hasStarred', false);
    case TOGGLE_SETTINGS_MODAL:
      return state.set('showSettingsModal', !state.get('showSettingsModal'));
    default:
      return state;
  }
};
