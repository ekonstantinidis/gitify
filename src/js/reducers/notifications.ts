import * as _ from 'lodash';
import {
  NOTIFICATIONS,
  MARK_NOTIFICATION,
  MARK_REPO_NOTIFICATION,
  UNSUBSCRIBE_NOTIFICATION,
} from '../actions';
import { LOGOUT } from '../../types/actions';
import { Notification } from '../../types/github';
import { NotificationsState } from '../../types/reducers';

const initialState: NotificationsState = {
  response: [],
  isFetching: false,
  failed: false,
};

export default function reducer(
  state = initialState,
  action
): NotificationsState {
  switch (action.type) {
    case NOTIFICATIONS.REQUEST:
      return { ...state, isFetching: true, failed: false };
    case NOTIFICATIONS.SUCCESS:
      return { ...state, isFetching: false, response: action.payload };
    case NOTIFICATIONS.FAILURE:
      return { ...state, isFetching: false, failed: true, response: [] };
    case MARK_NOTIFICATION.SUCCESS:
    case UNSUBSCRIBE_NOTIFICATION.SUCCESS:
      const accountIndex = state.response.findIndex(
        (obj) => obj.hostname === action.meta.hostname
      );

      return _.updateWith(
        { ...state },
        `[response][${accountIndex}][notifications]`,
        (accNotifications: Notification[]) => {
          return accNotifications.filter(
            (notification) => notification.id !== action.meta.id
          );
        }
      );
    case MARK_REPO_NOTIFICATION.SUCCESS:
      const accNotificationsRepoIndex = state.response.findIndex(
        (obj) => obj.hostname === action.meta.hostname
      );

      return _.updateWith(
        { ...state },
        `[response][${accNotificationsRepoIndex}][notifications]`,
        (accNotifications) => {
          return accNotifications.filter(
            (notification) =>
              notification.repository.full_name !== action.meta.repoSlug
          );
        }
      );
    case LOGOUT:
      return initialState;
    default:
      return state;
  }
}
