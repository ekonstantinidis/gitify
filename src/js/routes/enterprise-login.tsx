const ipcRenderer = require('electron').ipcRenderer;

import * as React from 'react';
import { Form, FormRenderProps } from 'react-final-form';
import { connect } from 'react-redux';
import { ArrowLeftIcon } from '@primer/octicons-react';

import { AppState } from '../../types/reducers';
import { authGithub } from '../utils/helpers';
import { FieldInput } from '../components/fields/input';

interface IValues {
  hostname?: string;
  clientId?: string;
  clientSecret?: string;
}

interface IFormErrors {
  hostname?: string;
  clientId?: string;
  clientSecret?: string;
}

export const validate = (values: IValues): IFormErrors => {
  const errors: IFormErrors = {};
  if (!values.hostname) {
    errors.hostname = 'Required';
  } else if (
    !/^([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}$/i.test(
      values.hostname
    )
  ) {
    errors.hostname = 'Invalid hostname.';
  }

  if (!values.clientId) {
    // 20
    errors.clientId = 'Required';
  } else if (!/^[A-Z0-9]{20}$/i.test(values.clientId)) {
    errors.clientId = 'Invalid client id.';
  }

  if (!values.clientSecret) {
    // 40
    errors.clientSecret = 'Required';
  } else if (!/^[A-Z0-9]{40}$/i.test(values.clientSecret)) {
    errors.clientSecret = 'Invalid client secret.';
  }

  return errors;
};

interface IProps {
  history: any;
  dispatch: any;
  enterpriseAccountsCount: number;
}

interface IState {
  enterpriseAccountsCount: number;
}

export class EnterpriseLogin extends React.Component<IProps, IState> {
  state = {
    enterpriseAccountsCount: 0,
  };

  static getDerivedStateFromProps(props: IProps, state) {
    if (props.enterpriseAccountsCount > state.enterpriseAccountsCount) {
      ipcRenderer.send('reopen-window');
      props.history.goBack();
    }

    return {
      enterpriseAccountsCount: props.enterpriseAccountsCount,
    };
  }

  renderForm = (formProps: FormRenderProps) => {
    const { handleSubmit, submitting, pristine } = formProps;

    return (
      <form onSubmit={handleSubmit}>
        <FieldInput
          name="hostname"
          label="Hostname"
          placeholder="github.company.com"
        />

        <FieldInput name="clientId" label="Client ID" placeholder="123456789" />

        <FieldInput
          name="clientSecret"
          label="Client Secret"
          placeholder="ABC123DEF456"
        />

        <button
          className="float-right px-4 py-2 mt-4 bg-gray-300 font-semibold rounded text-sm text-center hover:bg-gray-500 hover:text-white focus:outline-none"
          disabled={submitting || pristine}
          type="submit"
          title="Login Button"
        >
          <span>Login</span>
        </button>
      </form>
    );
  };

  handleSubmit(data, dispatch) {
    authGithub(data, dispatch);
  }

  render() {
    return (
      <div>
        <div className="flex justify-between items-center mt-4 py-2 mx-8">
          <button
            className="focus:outline-none"
            aria-label="Go Back"
            onClick={() => this.props.history.goBack()}
          >
            <ArrowLeftIcon size={20} className="hover:text-gray-400" />
          </button>

          <h3 className="text-lg font-semibold">
            Login with GitHub Enterprise
          </h3>
        </div>

        <div className="flex-1 px-8">
          <Form
            initialValues={{
              hostname: '',
              clientId: '',
              clientSecret: '',
            }}
            onSubmit={(data) => this.handleSubmit(data, this.props.dispatch)}
            validate={validate}
          >
            {this.renderForm}
          </Form>
        </div>
      </div>
    );
  }
}

export function mapStateToProps(state: AppState) {
  return {
    enterpriseAccountsCount: state.auth.enterpriseAccounts.length,
  };
}

export default connect(mapStateToProps, null)(EnterpriseLogin);
