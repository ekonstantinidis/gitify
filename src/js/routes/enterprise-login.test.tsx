import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { mockedEnterpriseAccounts } from '../__mocks__/mockedData';

const { ipcRenderer, remote } = require('electron');
const BrowserWindow = remote.BrowserWindow;
const dialog = remote.dialog;

import { EnterpriseLogin, mapStateToProps, validate } from './enterprise-login';
import { AppState } from '../../types/reducers';

describe('routes/enterprise-login.js', () => {
  const props = {
    enterpriseAccountsCount: 0,
    dispatch: jest.fn(),
    handleSubmit: (cb) => cb,
    history: {
      goBack: jest.fn(),
    },
  };

  beforeEach(function () {
    // @ts-ignore
    new BrowserWindow().loadURL.mockReset();
    spyOn(ipcRenderer, 'send');
    props.dispatch.mockReset();
  });

  it('should test the mapStateToProps method', () => {
    const state = {
      auth: {
        enterpriseAccounts: mockedEnterpriseAccounts,
      },
    } as AppState;

    const mappedProps = mapStateToProps(state);

    expect(mappedProps.enterpriseAccountsCount).toBe(
      mockedEnterpriseAccounts.length
    );
  });

  it('renders correctly', () => {
    const tree = TestRenderer.create(
      <Provider store={createStore(() => {})}>
        <MemoryRouter>
          <EnterpriseLogin {...props} />
        </MemoryRouter>
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });

  it('should validate the form values', () => {
    let values;
    const emptyValues = {
      hostname: null,
      clientId: null,
      clientSecret: null,
    };

    values = {
      ...emptyValues,
    };
    expect(validate(values).hostname).toBe('Required');
    expect(validate(values).clientId).toBe('Required');
    expect(validate(values).clientSecret).toBe('Required');

    values = {
      ...emptyValues,
      hostname: 'hello',
      clientId: '!@£INVALID-.1',
      clientSecret: '!@£INVALID-.1',
    };
    expect(validate(values).hostname).toBe('Invalid hostname.');
    expect(validate(values).clientId).toBe('Invalid client id.');
    expect(validate(values).clientSecret).toBe('Invalid client secret.');
  });

  it('should receive a logged-in enterprise account', () => {
    const caseProps = {
      ...props,
    };

    const { rerender } = render(
      <Provider store={createStore(() => {})}>
        <MemoryRouter>
          <EnterpriseLogin {...caseProps} />
        </MemoryRouter>
      </Provider>
    );

    rerender(
      <Provider store={createStore(() => {})}>
        <MemoryRouter>
          <EnterpriseLogin
            {...caseProps}
            enterpriseAccountsCount={props.enterpriseAccountsCount + 1}
          />
        </MemoryRouter>
      </Provider>
    );

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('reopen-window');
    expect(props.history.goBack).toHaveBeenCalledTimes(1);
  });

  it('should render the form with errors', () => {
    const { getByLabelText, getByTitle, getByText } = render(
      <Provider store={createStore(() => {})}>
        <MemoryRouter>
          <EnterpriseLogin {...props} />
        </MemoryRouter>
      </Provider>
    );

    fireEvent.change(getByLabelText('Hostname'), {
      target: { value: 'test' },
    });
    fireEvent.change(getByLabelText('Client ID'), {
      target: { value: '123' },
    });
    fireEvent.change(getByLabelText('Client Secret'), {
      target: { value: 'abc' },
    });

    fireEvent.submit(getByTitle('Login Button'));

    expect(getByText('Invalid hostname.')).toBeTruthy();
    expect(getByText('Invalid client id.')).toBeTruthy();
    expect(getByText('Invalid client secret.')).toBeTruthy();
  });

  it('should open the login window and get a code successfully (will-redirect)', () => {
    const code = '123123123';
    const hostname = 'github.gitify.io';
    const clientId = '12312312312312312312';

    spyOn(new BrowserWindow().webContents, 'on').and.callFake(
      (event, callback) => {
        if (event === 'will-redirect') {
          const event = new Event('will-redirect');
          callback(event, `https://${hostname}/?code=${code}`);
        }
      }
    );

    const expectedUrl = `https://${hostname}/login/oauth/authorize?client_id=${clientId}&scope=user:email,notifications`;

    const { getByLabelText, getByTitle } = render(
      <Provider store={createStore(() => {})}>
        <MemoryRouter>
          <EnterpriseLogin {...props} />
        </MemoryRouter>
      </Provider>
    );

    fireEvent.change(getByLabelText('Hostname'), {
      target: { value: hostname },
    });
    fireEvent.change(getByLabelText('Client ID'), {
      target: { value: clientId },
    });
    fireEvent.change(getByLabelText('Client Secret'), {
      target: { value: 'ABC123ABCDABC123ABCDABC123ABCDABC123ABCD' },
    });

    fireEvent.submit(getByTitle('Login Button'));

    expect(new BrowserWindow().loadURL).toHaveBeenCalledTimes(1);
    expect(new BrowserWindow().loadURL).toHaveBeenCalledWith(expectedUrl);
    expect(props.dispatch).toHaveBeenCalledTimes(1);
  });
});
