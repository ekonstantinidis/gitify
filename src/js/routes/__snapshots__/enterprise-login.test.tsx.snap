// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/enterprise-login.js renders correctly 1`] = `
<div>
  <div
    className="sc-Axmtr ebMSwT"
  >
    <h3
      className="sc-fzozJi jqfsXz"
    >
      Login to GitHub Enterprise
    </h3>
    <button
      aria-label="Go Back"
      className="sc-fzoLsD cUlvxj"
      onClick={[Function]}
    >
      Ã—
    </button>
  </div>
  <div
    className="sc-AxmLO hfTlrw"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="sc-AxjAm euTnqr"
      >
        <label
          className="sc-AxirZ iDECDC"
          htmlFor="hostname"
        >
          Hostname
        </label>
        <input
          className="sc-AxiKw gyFGja"
          id="hostname"
          name="hostname"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="github.company.com"
          type="text"
          value=""
        />
      </div>
      <div
        className="sc-AxjAm euTnqr"
      >
        <label
          className="sc-AxirZ iDECDC"
          htmlFor="clientId"
        >
          Client ID
        </label>
        <input
          className="sc-AxiKw gyFGja"
          id="clientId"
          name="clientId"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="123456789"
          type="text"
          value=""
        />
      </div>
      <div
        className="sc-AxjAm euTnqr"
      >
        <label
          className="sc-AxirZ iDECDC"
          htmlFor="clientSecret"
        >
          Client Secret
        </label>
        <input
          className="sc-AxiKw gyFGja"
          id="clientSecret"
          name="clientSecret"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="ABC123DEF456"
          type="text"
          value=""
        />
      </div>
      <button
        className="sc-AxheI dRrlwj"
        disabled={true}
        title="Login Button"
        type="submit"
      >
        <svg
          aria-hidden="true"
          className="octicon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path fill-rule=\\"evenodd\\" d=\\"M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0016 8c0-4.42-3.58-8-8-8z\\"></path>",
            }
          }
          fill="currentColor"
          height={16}
          role="img"
          style={
            Object {
              "display": "inline-block",
              "userSelect": "none",
              "verticalAlign": "text-bottom",
            }
          }
          viewBox="0 0 16 16"
          width={16}
        />
        <span>
          Login to GitHub Enterprise
        </span>
      </button>
    </form>
  </div>
</div>
`;
